kimchi/src/circuits/polynomials/xor.rs

```diff
@@ -0,0 +1,546 @@
+//! This module includes the definition of the XOR gadget for 64, 32, and 16 bits,
+//! the definition of the constraints of the `Xor16` circuit gate,
+//! and the code for witness generation for the XOR gadget.
+use crate::{
+    alphas::Alphas,
+    circuits::{
+        argument::{Argument, ArgumentEnv, ArgumentType},
+        constraints::ConstraintSystem,
+        expr::{self, constraints::ExprOps, l0_1, Environment, LookupEnvironment, E},
+        gate::{CircuitGate, CircuitGateError, CircuitGateResult, Connect, GateType},
+        lookup::{
+            self,
+            lookups::{LookupInfo, LookupsUsed},
+            tables::{GateLookupTable, LookupTable},
+        },
+        polynomial::COLUMNS,
+        wires::Wire,
+        witness::{self, ConstantCell, CopyBitsCell, CrumbCell, Variables, WitnessCell},
+    },
+    curve::KimchiCurve,
+    variable_map,
+};
+use ark_ff::{PrimeField, Zero};
+use ark_poly::{
+    univariate::DensePolynomial, EvaluationDomain, Evaluations, Radix2EvaluationDomain as D,
+};
+use rand::{rngs::StdRng, SeedableRng};
+use std::{array, collections::HashMap, marker::PhantomData};
+
+pub const GATE_COUNT: usize = 1;
+
+impl<F: PrimeField> CircuitGate<F> {
+    /// Creates a XOR gadget for `bits` length
+    /// Includes:
+    /// - num_xors Xor16 gates
+    /// - 1 Generic gate to constrain the final row to be zero with itself
+    /// Outputs tuple (next_row, circuit_gates) where
+    /// - next_row  : next row after this gate
+    /// - gates     : vector of circuit gates comprising this gate
+    pub fn create_xor(new_row: usize, bits: usize) -> (usize, Vec<Self>) {
+        let num_xors = num_xors(bits);
+        let mut gates = (0..num_xors)
+            .map(|i| CircuitGate {
+                typ: GateType::Xor16,
+                wires: Wire::new(new_row + i),
+                coeffs: vec![],
+            })
+            .collect::<Vec<_>>();
+        let zero_row = new_row + num_xors;
+        gates.push(CircuitGate {
+            typ: GateType::Generic,
+            wires: Wire::new(zero_row),
+            coeffs: vec![F::one(), F::one(), -F::one()],
```

Comment:
I don't think we want to check that `fin_in1 + fin_in2 - fin_out = 0`. Just constrain `fin_in = 0` with `vec[F::one()]`